File: src/db/evidence.py
from __future__ import annotations

import hashlib
import hmac
import inspect
import json
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any
from uuid import UUID

from sqlalchemy import DateTime, String, func, select
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Mapped, mapped_column

from src.db.connection import Base

GENESIS_PREV_HASH = "genesis"
EVIDENCE_CHAIN_LOCK_KEY = 704281913

PII_PAYLOAD_KEYS = {"user_id", "email", "account_ref", "wa_id"}


@dataclass(frozen=True, slots=True)
class EventSpec:
    """Metadata for one auditable event type."""

    description: str
    entity_type: str
    generates_receipt: bool = False
    public_fields: frozenset[str] = field(default_factory=frozenset)
    delayed_fields: frozenset[str] = field(default_factory=frozenset)


EVENT_CATALOG: dict[str, EventSpec] = {
    # --- Submission lifecycle ---
    "submission_received": EventSpec(
        description="Submission recorded",
        entity_type="submission",
        public_fields=frozenset({"submission_id", "language", "status", "hash", "raw_text", "reason_code"}),
    ),
    "submission_not_eligible": EventSpec(
        description="Submission rejected: user not eligible",
        entity_type="user",
        public_fields=frozenset({"reason_code"}),
    ),
    "submission_rate_limited": EventSpec(
        description="Submission rejected: rate limit",
        entity_type="user",
        public_fields=frozenset({"reason_code", "limit_type"}),
    ),
    "submission_rejected_not_policy": EventSpec(
        description="Submission rejected: not a policy proposal",
        entity_type="submission",
        public_fields=frozenset({"submission_id", "rejection_reason", "model_version", "prompt_version"}),
    ),
    # --- Candidate lifecycle ---
    "candidate_created": EventSpec(
        description="Policy candidate created from submission",
        entity_type="submission",
        public_fields=frozenset({
            "submission_id", "title", "summary", "stance",
            "policy_topic", "policy_key", "confidence", "model_version", "prompt_version",
        }),
    ),
    # --- Cluster lifecycle ---
    "cluster_created": EventSpec(
        description="New policy cluster formed",
        entity_type="cluster",
        public_fields=frozenset({"cluster_id", "policy_key", "policy_topic", "member_count"}),
    ),
    "cluster_updated": EventSpec(
        description="Cluster membership changed",
        entity_type="cluster",
        public_fields=frozenset({"cluster_id", "policy_key", "old_member_count", "new_member_count"}),
    ),
    "cluster_merged": EventSpec(
        description="Clusters merged during normalization",
        entity_type="cluster",
        public_fields=frozenset({"survivor_key", "merged_key", "merged_cluster_id", "new_member_count"}),
    ),
    "ballot_question_generated": EventSpec(
        description="Ballot question generated for cluster",
        entity_type="cluster",
        public_fields=frozenset({"policy_key", "ballot_question", "member_count", "model_version"}),
    ),
    "policy_options_generated": EventSpec(
        description="Stance options generated for cluster",
        entity_type="cluster",
        public_fields=frozenset({"cluster_id", "option_count", "labels", "model_version"}),
    ),
    # --- Endorsement lifecycle ---
    "policy_endorsed": EventSpec(
        description="User endorsed a policy cluster",
        entity_type="policy_endorsement",
        generates_receipt=True,
        public_fields=frozenset({"cluster_id"}),
    ),
    "endorsement_not_eligible": EventSpec(
        description="Endorsement rejected: user not eligible",
        entity_type="user",
        public_fields=frozenset({"reason_code"}),
    ),
    # --- Voting lifecycle ---
    "vote_cast": EventSpec(
        description="Vote recorded in active cycle",
        entity_type="vote",
        generates_receipt=True,
        public_fields=frozenset({"cycle_id"}),
        delayed_fields=frozenset({"approved_cluster_ids", "selections"}),
    ),
    "vote_not_eligible": EventSpec(
        description="Vote rejected: user not eligible",
        entity_type="user",
        public_fields=frozenset({"reason_code", "cycle_id"}),
    ),
    "vote_change_limit_reached": EventSpec(
        description="Vote change rejected: limit reached",
        entity_type="user",
        public_fields=frozenset({"cycle_id"}),
    ),
    "cycle_opened": EventSpec(
        description="Voting cycle opened",
        entity_type="voting_cycle",
        public_fields=frozenset({
            "cycle_id", "cluster_ids", "starts_at", "ends_at", "cycle_duration_hours",
        }),
    ),
    "cycle_closed": EventSpec(
        description="Voting cycle closed and tallied",
        entity_type="voting_cycle",
        public_fields=frozenset({"total_voters", "results"}),
    ),
    # --- Identity ---
    "user_verified": EventSpec(
        description="User identity verified",
        entity_type="user",
        public_fields=frozenset({"method"}),
    ),
    # --- Disputes ---
    "dispute_escalated": EventSpec(
        description="Dispute escalated to ensemble",
        entity_type="submission",
        public_fields=frozenset({
            "threshold", "primary_model", "primary_confidence",
            "ensemble_models", "selected_model", "selected_confidence",
        }),
    ),
    "dispute_resolved": EventSpec(
        description="Dispute resolved by automated review",
        entity_type="submission",
        public_fields=frozenset({
            "submission_id", "candidate_id", "escalated", "confidence",
            "model_version", "resolved_title", "resolved_summary", "resolution_seconds",
        }),
    ),
    "dispute_metrics_recorded": EventSpec(
        description="Dispute metrics checkpoint recorded",
        entity_type="system",
        public_fields=frozenset({"period", "total_disputes", "escalation_rate", "avg_resolution_seconds"}),
    ),
    "dispute_tuning_recommended": EventSpec(
        description="Model/policy tuning recommended based on dispute metrics",
        entity_type="system",
        public_fields=frozenset({"reason", "metrics"}),
    ),
    # --- Anchoring ---
    "anchor_computed": EventSpec(
        description="Daily Merkle root computed",
        entity_type="daily_anchor",
        public_fields=frozenset({"day", "merkle_root", "entry_count"}),
    ),
    "anchor_publish_attempted": EventSpec(
        description="Witness publication attempted",
        entity_type="daily_anchor",
        public_fields=frozenset({"day", "merkle_root"}),
    ),
    "anchor_publish_succeeded": EventSpec(
        description="Witness publication succeeded",
        entity_type="daily_anchor",
        public_fields=frozenset({"day", "merkle_root", "receipt"}),
    ),
    "anchor_publish_failed": EventSpec(
        description="Witness publication failed",
        entity_type="daily_anchor",
        public_fields=frozenset({"day", "merkle_root", "error_type"}),
    ),
    # --- Voice ---
    "voice_enrolled": EventSpec(
        description="Voice enrollment completed",
        entity_type="user",
        public_fields=frozenset({"phrase_count", "model_version"}),
    ),
    "voice_enroll_phrase_rejected": EventSpec(
        description="Voice enrollment phrase rejected",
        entity_type="user",
        public_fields=frozenset({"phrase_id", "reject_reason"}),
    ),
    "voice_verified": EventSpec(
        description="Voice verification completed",
        entity_type="user",
        public_fields=frozenset({"result", "phrase_id", "reject_reason"}),
    ),
}

VALID_EVENT_TYPES = set(EVENT_CATALOG.keys())


class EvidenceLogEntry(Base):
    __tablename__ = "evidence_log"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC)
    )
    event_type: Mapped[str] = mapped_column(String, nullable=False)
    entity_type: Mapped[str] = mapped_column(String, nullable=False)
    entity_id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), nullable=False, index=True)
    payload: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    hash: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    prev_hash: Mapped[str] = mapped_column(String(64), nullable=False)


def canonical_json(obj: dict[str, Any]) -> str:
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def isoformat_z(ts: datetime) -> str:
    dt = ts.astimezone(UTC)
    return dt.isoformat(timespec="milliseconds").replace("+00:00", "Z")


def compute_entry_hash(
    *,
    timestamp_iso: str,
    event_type: str,
    entity_type: str,
    entity_id: str,
    payload: dict[str, Any],
    prev_hash: str,
) -> str:
    material = {
        "timestamp": timestamp_iso,
        "event_type": event_type,
        "entity_type": entity_type,
        "entity_id": entity_id.lower(),
        "payload": payload,
        "prev_hash": prev_hash,
    }
    serialized = canonical_json(material)
    return hashlib.sha256(serialized.encode("utf-8")).hexdigest()


async def append_evidence(
    session: AsyncSession,
    event_type: str,
    entity_type: str,
    entity_id: UUID,
    payload: dict[str, Any],
) -> EvidenceLogEntry:
    if event_type not in VALID_EVENT_TYPES:
        raise ValueError(f"Invalid event_type: {event_type}")

    async with session.begin_nested():
        # Serialize append operations so concurrent writers cannot reuse the same prev_hash.
        bind = session.get_bind()
        if inspect.isawaitable(bind):
            bind = await bind
        dialect_name = getattr(getattr(bind, "dialect", None), "name", None)
        if dialect_name == "postgresql":
            await session.execute(select(func.pg_advisory_xact_lock(EVIDENCE_CHAIN_LOCK_KEY)))

        last_result = await session.execute(
            select(EvidenceLogEntry).order_by(EvidenceLogEntry.id.desc()).limit(1).with_for_update()
        )
        last_entry = last_result.scalar_one_or_none()
        prev_hash = last_entry.hash if last_entry else GENESIS_PREV_HASH
        timestamp = datetime.now(UTC)
        entry_hash = compute_entry_hash(
            timestamp_iso=isoformat_z(timestamp),
            event_type=event_type,
            entity_type=entity_type,
            entity_id=str(entity_id),
            payload=payload,
            prev_hash=prev_hash,
        )
        entry = EvidenceLogEntry(
            timestamp=timestamp,
            event_type=event_type,
            entity_type=entity_type,
            entity_id=entity_id,
            payload=payload,
            hash=entry_hash,
            prev_hash=prev_hash,
        )
        session.add(entry)
        await session.flush()
        await session.refresh(entry)
        return entry


async def verify_chain(session: AsyncSession) -> tuple[bool, int]:
    result = await session.execute(select(EvidenceLogEntry).order_by(EvidenceLogEntry.id.asc()))
    entries = list(result.scalars().all())
    prev_hash = GENESIS_PREV_HASH

    for index, entry in enumerate(entries):
        expected = compute_entry_hash(
            timestamp_iso=isoformat_z(entry.timestamp),
            event_type=entry.event_type,
            entity_type=entry.entity_type,
            entity_id=str(entry.entity_id),
            payload=entry.payload,
            prev_hash=entry.prev_hash,
        )
        if entry.hash != expected:
            return (False, index + 1)
        if entry.prev_hash != prev_hash:
            return (False, index + 1)
        prev_hash = entry.hash

    return (True, len(entries))


def generate_receipt_token(entry_hash: str, signing_key: str) -> str:
    """Create an HMAC receipt token binding an evidence entry to a signing key.

    Users receive this token as proof their action was included in the chain.
    Verification is stateless: recompute HMAC and compare.
    """
    return hmac.new(
        signing_key.encode("utf-8"), entry_hash.encode("utf-8"), hashlib.sha256
    ).hexdigest()


def verify_receipt_token(entry_hash: str, signing_key: str, token: str) -> bool:
    expected = generate_receipt_token(entry_hash, signing_key)
    return hmac.compare_digest(expected, token)


def strip_evidence_pii(payload: dict[str, Any]) -> dict[str, Any]:
    """Recursively strip PII keys from evidence payload for public display."""
    result: dict[str, Any] = {}
    for k, v in payload.items():
        if k in PII_PAYLOAD_KEYS:
            continue
        if isinstance(v, dict):
            result[k] = strip_evidence_pii(v)
        elif isinstance(v, list):
            result[k] = [strip_evidence_pii(item) if isinstance(item, dict) else item for item in v]
        else:
            result[k] = v
    return result


def apply_visibility_tier(
    event_type: str,
    payload: dict[str, Any],
    *,
    cycle_closed: bool = False,
) -> dict[str, Any]:
    """Apply visibility-tier filtering on top of PII stripping.

    - Strips PII recursively.
    - Removes delayed fields when cycle is still active.
    """
    cleaned = strip_evidence_pii(payload)
    spec = EVENT_CATALOG.get(event_type)
    if spec and spec.delayed_fields and not cycle_closed:
        cleaned = {k: v for k, v in cleaned.items() if k not in spec.delayed_fields}
    return cleaned

File: web/lib/evidence.ts
export type EvidenceEntry = {
  id: number;
  timestamp: string;
  event_type: string;
  entity_type: string;
  entity_id: string;
  payload: Record<string, unknown>;
  hash: string;
  prev_hash: string;
};

export type EvidenceResponse = {
  total: number;
  page: number;
  per_page: number;
  entries: EvidenceEntry[];
};

async function sha256Hex(value: string): Promise<string> {
  const data = new TextEncoder().encode(value);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Matches Python's json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False).
 * Recursively sorts object keys at every nesting level.
 */
export function canonicalJson(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "number" || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return "[" + value.map(canonicalJson).join(",") + "]";
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  const parts = keys.map((k) => JSON.stringify(k) + ":" + canonicalJson(obj[k]));
  return "{" + parts.join(",") + "}";
}

export async function verifyChain(entries: EvidenceEntry[]): Promise<{valid: boolean; failedIndex?: number}> {
  let previous = "genesis";
  for (let index = 0; index < entries.length; index += 1) {
    const entry = entries[index];
    const material = {
      timestamp: entry.timestamp,
      event_type: entry.event_type,
      entity_type: entry.entity_type,
      entity_id: entry.entity_id.toLowerCase(),
      payload: entry.payload,
      prev_hash: entry.prev_hash
    };
    const serialized = canonicalJson(material);
    const expected = await sha256Hex(serialized);
    if (entry.hash !== expected || entry.prev_hash !== previous) {
      return {valid: false, failedIndex: index};
    }
    previous = entry.hash;
  }
  return {valid: true};
}

const DELIBERATION_EVENT_TYPES = new Set([
  "submission_received",
  "submission_not_eligible",
  "submission_rate_limited",
  "submission_rejected_not_policy",
  "candidate_created",
  "cluster_created",
  "cluster_updated",
  "cluster_merged",
  "ballot_question_generated",
  "policy_options_generated",
  "vote_cast",
  "vote_not_eligible",
  "vote_change_limit_reached",
  "policy_endorsed",
  "endorsement_not_eligible",
  "cycle_opened",
  "cycle_closed",
  "dispute_escalated",
  "dispute_resolved",
]);

export function isDeliberationEvent(eventType: string): boolean {
  return DELIBERATION_EVENT_TYPES.has(eventType);
}

export type FilterCategory = "submissions" | "policies" | "votes" | "disputes" | "users" | "system";

export const EVENT_CATEGORIES: Record<FilterCategory, string[]> = {
  submissions: ["submission_received", "submission_not_eligible", "submission_rate_limited", "submission_rejected_not_policy"],
  policies: ["candidate_created", "cluster_created", "cluster_updated", "cluster_merged", "ballot_question_generated", "policy_options_generated"],
  votes: ["vote_cast", "vote_not_eligible", "vote_change_limit_reached", "policy_endorsed", "endorsement_not_eligible", "cycle_opened", "cycle_closed"],
  disputes: ["dispute_escalated", "dispute_resolved"],
  users: ["user_verified"],
  system: ["anchor_computed", "anchor_publish_attempted", "anchor_publish_succeeded", "anchor_publish_failed", "dispute_metrics_recorded", "dispute_tuning_recommended"],
};

function truncate(text: string, max: number): string {
  if (text.length <= max) return text;
  return text.slice(0, max) + "...";
}

function str(v: unknown): string {
  return typeof v === "string" ? v : String(v ?? "");
}

export function eventDescription(
  entry: EvidenceEntry,
  t: (key: string, values?: Record<string, string | number>) => string,
): string {
  const p = entry.payload;
  switch (entry.event_type) {
    case "submission_received": {
      if (p.status === "rejected_high_risk_pii") return t("events.submissionRejectedPii");
      const text = str(p.raw_text);
      return text
        ? t("events.submissionReceived", {text: truncate(text, 80)})
        : t("events.submissionReceivedGeneric");
    }
    case "submission_not_eligible":
      return t("events.submissionNotEligible");
    case "submission_rate_limited":
      return t("events.submissionRateLimited");
    case "submission_rejected_not_policy":
      return t("events.submissionRejectedNotPolicy");
    case "candidate_created":
      return t("events.candidateCreated", {
        title: truncate(str(p.title), 60),
        topic: str(p.policy_topic),
        confidence: String(Math.round(Number(p.confidence ?? 0) * 100)),
      });
    case "cluster_created":
      return t("events.clusterCreated", {
        summary: truncate(str(p.summary ?? p.policy_key ?? ""), 60),
        memberCount: String(p.member_count ?? "?"),
      });
    case "cluster_updated":
      return t("events.clusterUpdated", {
        summary: truncate(str(p.summary ?? p.policy_key ?? ""), 60),
        memberCount: String(p.new_member_count ?? p.member_count ?? "?"),
      });
    case "cluster_merged":
      return t("events.clusterMerged", {
        mergedKey: str(p.merged_key),
        survivorKey: str(p.survivor_key),
      });
    case "ballot_question_generated":
      return t("events.ballotQuestionGenerated", {
        policyKey: str(p.policy_key),
      });
    case "policy_options_generated":
      return t("events.policyOptionsGenerated", {
        optionCount: String(p.option_count ?? "?"),
      });
    case "vote_cast":
      return t("events.voteCast", {
        count: String((p.approved_cluster_ids as unknown[])?.length ?? 0),
      });
    case "vote_not_eligible":
      return t("events.voteNotEligible");
    case "vote_change_limit_reached":
      return t("events.voteChangeLimitReached");
    case "policy_endorsed":
      return t("events.policyEndorsed");
    case "endorsement_not_eligible":
      return t("events.endorsementNotEligible");
    case "cycle_opened":
      return t("events.cycleOpened", {
        count: String((p.cluster_ids as unknown[])?.length ?? 0),
      });
    case "cycle_closed":
      return t("events.cycleClosed", {
        totalVoters: String(p.total_voters ?? 0),
      });
    case "dispute_escalated":
      return t("events.disputeEscalated");
    case "dispute_resolved":
      return p.resolved_title
        ? t("events.disputeResolved", {title: truncate(str(p.resolved_title), 60)})
        : t("events.disputeResolvedGeneric");
    case "user_verified":
      return t("events.userVerified", {method: str(p.method)});
    case "anchor_computed":
      return t("events.anchorComputed", {
        entryCount: String(p.entry_count ?? 0),
      });
    case "anchor_publish_attempted":
      return t("events.anchorPublishAttempted");
    case "anchor_publish_succeeded":
      return t("events.anchorPublishSucceeded");
    case "anchor_publish_failed":
      return t("events.anchorPublishFailed");
    case "dispute_metrics_recorded":
      return t("events.disputeMetrics");
    case "dispute_tuning_recommended":
      return t("events.disputeTuning");
    default:
      return entry.event_type;
  }
}

export function entityLink(entry: EvidenceEntry, locale: string): string | null {
  switch (entry.entity_type) {
    case "cluster":
      return `/${locale}/collective-concerns/clusters/${entry.entity_id}`;
    case "voting_cycle":
      return `/${locale}/collective-concerns/community-votes`;
    case "vote":
      return `/${locale}/collective-concerns/community-votes`;
    default:
      return null;
  }
}

export function relativeTime(isoTimestamp: string): string {
  const now = Date.now();
  const then = new Date(isoTimestamp).getTime();
  const diffMs = now - then;
  const diffMin = Math.floor(diffMs / 60_000);
  if (diffMin < 1) return "just now";
  if (diffMin < 60) return `${diffMin}m ago`;
  const diffHours = Math.floor(diffMin / 60);
  if (diffHours < 24) return `${diffHours}h ago`;
  const diffDays = Math.floor(diffHours / 24);
  return `${diffDays}d ago`;
}
